src/DurableTask.Netherite/StorageLayer/Faster/AzureBlobs/AzureStorageDevice.cs

```diff
@@ -516,31 +556,83 @@ async Task ReadFromBlobAsync(UnmanagedMemoryStream stream, BlobUtilsV12.PageBlob
-                            }
+                            stream.Seek(offset, SeekOrigin.Begin);
 
-                            if (length > 0)
-                            {
-                                var client = (numAttempts > 1 || length == MAX_DOWNLOAD_SIZE) ? blob.Default : blob.Aggressive;
+                            long requestedLength = length;
 
-                                var response = await client.DownloadStreamingAsync(
-                                    range: new Azure.HttpRange(sourceAddress + offset, length),
-                                    conditions: null,
-                                    rangeGetContentHash: false,
-                                    cancellationToken: this.PartitionErrorHandler.Token)
-                                    .ConfigureAwait(false);
+                            if (readCap.HasValue && sourceAddress + offset + requestedLength > readCap.Value)
+                            {
+                                requestedLength = readCap.Value - (sourceAddress + offset);
 
-                                using (var streamingResult = response.Value)
+                                if (requestedLength <= 0)
                                 {
-                                    await streamingResult.Content.CopyToAsync(stream).ConfigureAwait(false);
+                                    requestedLength = 0;
                                 }
                             }
 
-                            if (stream.Position != offset + length)
+                            if (requestedLength > 0)
                             {
-                                throw new InvalidDataException($"wrong amount of data received from page blob, expected={length}, actual={stream.Position}");
+                                var client = (numAttempts > 1 || requestedLength == MAX_DOWNLOAD_SIZE) ? blob.Default : blob.Aggressive;
+
+                                try
+                                {
+                                    var response = await client.DownloadStreamingAsync(
+                                        range: new Azure.HttpRange(sourceAddress + offset, requestedLength),
+                                        conditions: null,
+                                        rangeGetContentHash: false,
+                                        cancellationToken: this.PartitionErrorHandler.Token)
+                                        .ConfigureAwait(false);
+
+                                    using (var streamingResult = response.Value)
+                                    {
+                                        await streamingResult.Content.CopyToAsync(stream).ConfigureAwait(false);
+                                    }
+
+                                    // we have observed that we may get 206 responses where the actual length is less than the requested length
+                                    // this seems to only happen if blob was enlarged in the past
```

Comment:
So what are we doing if we get a 206? I actually don't see us checking the response status explicitly, should we?
nit: let's also add in parenthesis the meaning of 206: "partial response", right?

Reply:
The SDK internally takes care of interpreting the status response when I access the streaming response (`response.Value`). From my perspective, what I see in this case is that the stream returned has fewer bytes than requested.

BTW, I have seen a few 206 during testing early on but later no longer saw them. I figured it makes sense to handle them just in case they pop up. It is pretty easy to handle them by just adjusting the next portion to start where the 206 ended.